### ISSUE TITLE
Replace GoogleAnalytics dependency with playservices ads identifier

### ISSUE THREAD

**Issue by plastiv** [adjust/android_sdk#366]
Adjust setup requires to add google analytics dependency
https://github.com/adjust/android_sdk#sdk-gps

These days Google is pushing android developers to use Firebase Analytics instead. Is it safe to replace `com.google.android.gms:play-services-analytics` with `com.google.android.gms:play-services-ads-identifier`? Some guidance is appreciated 🙏  

Since `AdvertisingIdClient` is a must dependency for adjust wouldn't it make more sense to depend on it directly and avoid developers to resolve it? Since dependency is implicit right now and there is no error if `AdvertisingIdClient` is missing on classpath.

**Comment by uerceg** [comment]
Hi @plastiv 

And thanks for the question, nice one. What we have noticed when testing a while ago was that even though Google has changed the location of the `AdvertisingIdClient` class to `-ads-identifier` package (if I recall well they moved it away from analytics somewhere around 15.0.0 version or something like that), usage of `-analytics` part seemed still to somehow automagically make that class present in the app. This is the reason why this is still in the README, but definitely something we were considering to change since we did the tests with `-ads-identifier` package and also worked just fine since these days that's where the `AdvertisingIdClient` lives. So I would say that in case you are really bothered for some reason with `-analytics` part, feel free to switch to `-ads-identifier`, it should not make any difference.

Why don't we depend on this API directly, also valid question. Lemme explain what is happening behind the scenes. Our SDK uses Google Play Services library as a weak dependency which, **if pinged** is pinged via reflection. There are two reasons for that decision:

1. Our SDK is aimed to be as generic as possible and is not exclusively used only by people who are publishing their apps to Google Play Store. Making dependency hard would definitely make our and Play Store developers' lives easier, but it would require for example clients that publish apps on stores which are unaware of Google Play Services and Google Advertising ID existence at all to have this dependency in their app for (for them) no obvious reason. This then leads to increase of app size, DEX method count, etc.
2. Our SDK uses reflection ping to Google Play Services library only as a fallback mechanism when trying to access information about advertising identifier and whether tracking is enabled or not. Primary way of reading it direct ping to Play Store app which requires no additional dependency and which is providing us with this info in huge majority of the cases.

With number 2 reason in mind, it is actually even hard for you to check if change of packages made any difference because most probably our primary way of obtaining information about advertising ID will get this info and fallback might never happen. In case you are really curious to test this (but not sure how many times would you need to test this before primary way fails) you can try to track all bunch of events and with each tracked event try to see the value of `gps_adid_src` parameter. If primary way in waterfall read it, value of this parameter will say `service`. If fallback mechanism took place, it will say `library`.

We also don't have any explicit log line which says anything about whether library is properly integrated or not is due to the fact that we would need to invoke reflection to do this check just for sake of providing this information and some clients are not happy when reflection calls take places and slow down things because (for them) not really justified reasons.

In case you have any other questions on this topic, feel free to ping us.

Cheers

**Comment by uerceg** [comment]
Feel free to reopen the issue in case you have any other questions on this topic. Cheers!

**Comment by TomorrowAndroid** [comment]
What is the current status of this?
Why have you still not updated the readme to use the `-ads-identifier` package?
Is there any other reason you still recommend `analytics` package?

We really would like to keep google analytics out of our app.

**Comment by uerceg** [comment]
Hi @TomorrowAndroid 

In theory, reasons are mostly historical since prior to Google's refactoring and reorganisation of Google Play Services library to packages, `-analytics` part was the one in charge for this functionality since it contained class we needed for this. Theoretically, that package works even today so if someone is using ancient version of Google Play Services, `-analytics` is still common denominator for everyone.

Today classes we need reside in `-ads-identifier` package, true. And also feel free to use just that part in your app since that's pretty much all you need. I guess enough time has passed and we should probably consider updating README to usage of this package only and for those using ancient version, point them out to update dependencies to latest ones.

Android SDK patch update is behind the corner, we'll add this change to README as well.

**Comment by calvarez-ov** [comment]
I didn't completely understand the detailed explanation.

* We don't have google analytics
* We currently have `com.google.android.gms:play-services-ads-identifier`.  Since our code never calls any of that library's apis directly, I was wondering if we could remove it.  I did some tests and it seems like Adjust is still able to get the advertising id even without this dependency in our project. `Adjust.getGoogleAdId()` still provides a value even without analytics or ads-identifier present in the app dependencies.  I see `gps_adid_src     service` in either case.


To sum up, what is the impact if we remove `ads-identifier` library, and we don't have analytics either?

Thanks!

**Comment by uerceg** [comment]
Hey @calvarez-ov 

And sorry for delay in replying in here. I'd say that your summary from comment above is pretty accurate. Having this dependency is useful only as fallback mechanism in case service approach we use to communicate with Play Store directly to get the advertising ID fails. But even that approach under the hood does pretty much the same as we do in our primary attempt. From what we saw, fallback happens extremely rarely.

So I'd say that it's probably fine to just keep that dependency out of your app. We will also check this topic more in detail for our next major SDK release and if it's really not needed (nowadays also kinda mostly legacy integration steps), we might completely remove that integration step.

Cheers!
